gobblin-core/src/main/java/gobblin/commit/FsRenameCommitStep.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.commit;
+
+import java.io.IOException;
+import java.net.URI;
+
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import com.google.common.base.Preconditions;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.util.HadoopUtils;
+
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * A {@link CommitStep} for renaming files within a {@link FileSystem} or between {@link FileSystem}s.
+ *
+ * @author ziliu
+ */
+@Slf4j
+public class FsRenameCommitStep extends CommitStepBase {
+
+  private final Path srcPath;
+  private final Path dstPath;
+  private final String srcFsUri;
+  private final String dstFsUri;
+  private transient FileSystem srcFs;
+  private transient FileSystem dstFs;
+
+  private FsRenameCommitStep(Builder builder) throws IOException {
+    super(builder);
+
+    Preconditions.checkNotNull(builder.srcPath);
+    Preconditions.checkNotNull(builder.dstPath);
+
+    this.srcPath = builder.srcPath;
+    this.dstPath = builder.dstPath;
+    this.srcFs = builder.srcFs != null ? builder.srcFs
+        : getFileSystem(this.props.getProp(ConfigurationKeys.FS_URI_KEY, ConfigurationKeys.LOCAL_FS_URI));
+    this.srcFsUri = this.srcFs.getUri().toString();
+    this.dstFs = builder.dstFs != null ? builder.dstFs
+        : getFileSystem(this.props.getProp(ConfigurationKeys.FS_URI_KEY, ConfigurationKeys.LOCAL_FS_URI));
+    this.dstFsUri = this.dstFs.getUri().toString();
+  }
+
+  public static class Builder extends CommitStepBase.Builder<Builder> {
+
+    public Builder() {
+      super();
+    }
+
+    public Builder(CommitSequence.Builder commitSequenceBuilder) {
+      super(commitSequenceBuilder);
+    }
+
+    private Path srcPath;
+    private Path dstPath;
+    private FileSystem srcFs;
+    private FileSystem dstFs;
+
+    public Builder from(Path srcPath) {
+      this.srcPath = srcPath;
+      return this;
+    }
+
+    public Builder to(Path dstPath) {
```

Comment:
Sorry, you're right, this will not produce that compile error because `withProps` is returning the correct type. However, this Builder is not easily extensible (for example, for a subclass of `FsRenameCommitStep`), will that be a problem?